_docs/cardano/2017-01-06-updating.md

```diff
@@ -157,93 +154,96 @@ modifications to the protocol with some degree of concurrency.
+would be polluted with meaningless keys. We would have to accept it, and that
+would bloat the blockchain.
+
+This is the motivation for the logic described below.
+
+In our implementation, block version can be in the following states:
+
+* **Adopted**, when the softfork resolution rule (see below) for a confirmed
+  block version is triggered.
+* **Confirmed**, when there is an update proposal that contains a confirmed
+  version of the software and this block version. Note that "confirmed version of
+  the software" is a technical term defined elsewhere. If there are multiple block
+  versions such that corresponding software is confirmed, but these versions
+  aren't adopted, we call them **competing**. For instance, there might be
+  versions `2.0.0`, `2.0.1`, `1.2.0`, `1.2.1`, `1.1.1` and `1.1.2`, with the last
+  adopted version being `1.1.3`. In this case the versions that are competing are
+  `2.0.0`, `2.0.1`, `1.2.0` and `1.2.1`. The older versions `1.1.1` and `1.1.2`
+  aren't competing, because `1.1.3` is already adopted.
+* Everything else. For example, a new block version is proposed, but the
+  software version isn't confirmed yet. There is no special name for this state.
+
+The softfork resolution rule works as follows:
+
+* Informally, a block version becomes *adopted* when a certain percentage of
+  stake (75% in the current implementation, but this may be different in the
+  mainnet) create a block with version `X`.
+* Formally, we do the following. First, recall that by design, our system does
+  not allow rollbacks of blockchain deeper than a certain fixed global threshold
+  `k`, which allows to define *stable* stake for each stakeholder as their stake
+  `k` blocks ago. When we process genesis block for epoch `e`, we compute the
+  stable stake of all leaders of all slots from the very beginning of the
+  network's existence. For each block version `X` that is currently *competing*,
+  we take all created *stable* blocks with version `X`, take set of all leaders of
+  these blocks and accumulate their stakes. If one of versions has ≥ 75%, it
+  gets *adopted*. If more than one version has ≥ 75%, we take one of them
+  deterministically.
+
+Note that new block versions can be adopted only at the beginning of an epoch,
+so all blocks in an epoch have the same block version.
 
 So, gathering everything up:
- 1. Once the update is approved, the protocol version (say v0.5) can be used.
- 2. Before a soft fork is resolved (i.e. the resolution rule hit), the
-   updated nodes should issue blocks with the new version v5, but treat
-   and validate blocks of version v5 as blocks of version v4.
- 3. Non-updated updated nodes accept new blocks:
-   a. Before the soft fork is resolved, discarding as per v4.
-   b. After the soft fork is resolved, as per v5 (i.e. allowing a block to
-  have unknown attributes, and so on).
+
+1. Once the update is approved, the protocol version (say `0.5.0`) can be used.
```

Comment:
That's a part from the previous version that I haven't edited yet, except for fixing the version numbers.